### ISSUE TITLE
Suggestion: Add new command line parameter `-localproxy`

### ISSUE THREAD

**Issue by lars18th** [cyd01/KiTTY#84]
Hi @cyd01 ,

In order to use the Cygtermd workaround with the command line, it's necessary to add a new parameter. The idea is quite similar to the old `-cygterm`, but now with the name `-localproxy`.

Example:
- `kitty_portable.exe -localproxy C:\Cygwin\bin\cygtermd.exe /home/user /bin/bash -login`

In this example the `-localproxy` parameter uses the path to run the `cygtermd.exe` binary using the "Telnet command, or local proxy command" option, **and** it enables the use of the "Consider proxying local host connections". Futhermore, it also uses a fake name for the target "Host name" (not used with cygtermd.exe but required for the session).

With this will be possible to overcome the Cygterm patch forever.
I hope you want to consider it.
Regards.

**Comment by lars18th** [comment]
Hi @cyd01 ,

The idea with this new command line parameter is to overcome the MinTTY of Cygwin at all. With this example it will be possible to open the KiTTY Window with the Cygwin Bash Shell in _any_ subdirectory (for example, when using the command "OPEN HERE..." from Windows):

`kitty_portable.exe -localproxy C:\Cygwin\bin\cygtermd.exe $(cygpath -u 'C:\Windows') /bin/bash -login`

So, please evaluate how to implement it.
Thank you!

**Comment by cyd01** [comment]
Hi
I've just implemented your first request.
There is now a `-localproxy` command-line parameter. Get this [beta build](https://www.9bis.net/kitty/files/kitty-debug.exe).
Here is the command line:
```
kitty.exe -localproxy 'C:\cygwin64\bin\cygtermd.exe /home/cyd /bin/bash -login' localhost
```
Of course adapt the **cygtermd** path to your configuration. It is mandatory to provide a **hostname target** (here `localhost`) even if it not used by **KiTTY** in this case. Without it the session can't start.
I was not able to implement your `$(cygpath -u 'C:\Windows')` feature. It seems the $(...) may be evaluated **before** the **Cygwin** environment starts.

**Comment by lars18th** [comment]
Hi @cyd01 ,

Thank you for the beta. However, it doesn't work for me. Every time it prints `unkown option "-login"` when no using quotes and when using single quotes. And with double quotes it opens a non-modal window with the shell. This is on Windows 7 64bit.

And regarding the `$(cygpath -u 'C:\Windows')`. The idea is to _evaluate_ it **before** the Cygwin environment starts. This tool converts from/to Windows<->Cygwin PATH syntax. So, the objective is that you provide a Windows directory path to start here, and the cygtermd process receives the same directory with the Cygwin syntax. So please, try to execute this in your CMD shell: `C:\cygwin64\bin\cygpath.exe C:\Windows`. It's the result `/cygdrive/c/Windows` ? Then it works ! So, the syntax for starting the KiTTY+Cygtermd will be `kitty.exe -localproxy 'C:\cygwin64\bin\cygtermd.exe /cygdrive/c/Windows /bin/bash -login' localhost` _after_ do the translation from `C:\Windows`.

Regards.

**Comment by cyd01** [comment]
It's what I've said.
$(...) is a Linux feature. It can't be interpreted by Windows. and of
course it works if you replace it by the Cygwin style directory
/cygdrive/c/....
So that you have to use C:\cygwin64\bin\cygpath.exe by yourself to build
the kitty.exe command line options.

I've tested it into my MinGW shell (where I compile hte beta builds). And
the quotes works fine on it. I will search for the equivalent for cmd.exe.

**Comment by lars18th** [comment]
Hi @cyd01 

> It's what I've said.
> $(...) is a Linux feature.

Yes, but...

> It can't be interpreted by Windows

The tool `cygpath.exe` can be called from Windows _and_ from Unix (Cygwin). So you can use it inside CMD _and_ inside Bash. The conversion is bidirectional, so you can convert from one format to another.

The problem is how to execute it **before** evaluate the command line parameters.
I'll try to think on ideas. Stay on hold.

> I've tested it into my MinGW shell (where I compile hte beta builds). And
> the quotes works fine on it. I will search for the equivalent for cmd.exe.

In Windows CMD quotes are everytime `"` (double). Let me some minutes to check something I missed.

**Comment by lars18th** [comment]
Hi @cyd01 ,

Regarding the quotes: They work! You only need to use double quotes.

Regarding my launch error: The problem is that this `kitty-debug.exe` executed inside an empty directory and without anything in the Registry, in this case it opens a window without decoration. I detected this in my computer as _everytime_ I use it in PORTABLE mode. Please, try to open this binary in a empty NewFolder and without data in the registry. Open a new session (manually) and you'll see the problem.

And finally, regarding the starting path: You think it will be a good solution to implement a patch for the  `cygtermd.exe` tool that uses a Windows path, instead or a Cygwin path?

**Comment by lars18th** [comment]
Hi @cyd01 ,

Regarding the enhancement of the `-localproxy` from my point of view you can merge the change. It really works! :+1: 

Only to comment (and document it):

- Use double quotes (`"`) to pass the value of the `localproxy` parameter.
- If you want to open the shell in some Windows path, then use:
```
kitty.exe -localproxy "C:\cygwin\bin\cygtermd.exe /bin/bash /bin/xhere /bin/bash 'C:\Windows'" localhost
```

The `/bin/xhere` is a script from the package `chere` that is created to launch a shell with the starting directory listed in the Windows path form. The only trick here is to launch before a bash shell that executes the xhere script. Then you can replace the `'C:\Windows'` with `.` or leave blank to open in the running directory (`"C:\cygwin\bin\cygtermd.exe /bin/bash /bin/xhere /bin/bash"` is valid).

Other than that, I recommend to remove all traces of the ancient cygterm patches in the _current_ code. It wil be more clear. You agree?
Good job! :wink:

**Comment by cyd01** [comment]
> 
> 
> Hi @cyd01 ,
> 
> Regarding the enhancement of the `-localproxy` from my point of view you can merge the change. It really works! 👍
> 
> Only to comment (and document it):
> 
>     * Use double quotes (`"`) to pass the value of the `localproxy` parameter.
> 
>     * If you want to open the shell in some Windows path, then use:
> 
> 
> ```
> kitty.exe -localproxy "C:\cygwin\bin\cygtermd.exe /bin/bash /bin/xhere /bin/bash 'C:\Windows'" localhost
> ```
> 
> The `/bin/xhere` is a script from the package `chere` that is created to launch a shell with the starting directory listed in the Windows path form. The only trick here is to launch before a bash shell that executes the xhere script. Then you can replace the `'C:\Windows'` with `.` or leave blank to open in the running directory (`"C:\cygwin\bin\cygtermd.exe /bin/bash /bin/xhere /bin/bash"` is valid).
> 
> Other than that, I recommend to remove all traces of the ancient cygterm patches in the _current_ code. It wil be more clear. You agree?
> Good job! 😉

I agree with, I'll remove old cygterm patch after merge.

**Comment by lars18th** [comment]
Implemented in v0.73.1.5, so I close this issue.

Anyway, please @cyd01 remember to remove all ancient cygterm code if you doesn't do it. :wink:

**Comment by johnwmail** [comment]
How to save this "
    kitty_portable.exe -localproxy C:\Cygwin\bin\cygtermd.exe /home/user /bin/bash -login
" in the GUI session ?
Thanks.
